text/3410-explicit-move-binding-mode.md

```diff
@@ -0,0 +1,265 @@
+immutable one, and this is not what `mut` does—it sets the binding mode to a mutable move.
+
+```rust
+let mut x_and_y: (i32, i32) = (25, -4);
+
+let (x, mut y) = &mut x_and_y;
+// The type of `x` is `&mut i32` and
+// the type of `y` is `i32` (and the binding is mutable)
+
+*x += 2; // `x_and_y` is modified
+y += 2; // `x_and_y` is not modified
+
+let (move x, y) = &x_and_y;
+// The type of `x` is `i32` and
+// the type of `y` is `&i32`
+```
+
+You can also switch to an immutable reference binding mode by using the `ref` keyword.
+
+```rust
+let mut x_y_z: (i32, i32, i32) = (400, 1, -99);
+let (x, ref y, move z) = &mut x_y_z;
+// The type of `x` is `&mut i32`,
+// the type of `y` is `&i32` and
+// the type of `z` is `i32`
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The syntax for _IdentifierPattern_ is updated as follows:
+> _IdentifierPattern_:
+> 
+> (`ref` | `move`)? `mut`? IDENTIFIER (`@` _PatternNoTopAlt_)?
+
+The binding mode of a binding depends on the default mode and
+the binding mode specifier (`mut`, `move`, `ref`, or `ref mut`)
+and is described by the following table. If the entry into the table is 
+followed by an exclamation mark in parentheses, a lint is triggered.
+The symbol “-//-” indicates that the entry is the same as the entry to the left,
+excluding whether it triggers the lint ((!)).
+
+| ↓specifier | →default = move       | reference | mutable reference |
+|------------|-----------------------|-----------|-------------------|
+| `mut`      | move mutable          | -//-      | -//-              |
+| `ref mut`  | mutable reference     | -//-      | -//- (!)          |
+| `ref`      | reference             | -//- (!)  | -//-              |
+| `move`     | move (!)              | -//-      | -//-              |
+| _none_     | move                  | reference | mutable reference |
+
+The lint is controlled by `unnecessary_binding_mode`. It is warn-by-default.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It can be argued that use of the `move` keyword should be replaced with
+  use of the `ref` keyword and not using match ergonomics at all.
+- This further entrenches the unintuitive fact that `mut` sets the binding
+  mode to a mutable move and disables referencing.
+- This means that either the grammar gets more complicated or an additional
+  sequence is allowed by the grammar but disallowed by the compiler.
+- The `move` keyword can be confusing here because a copy may happen instead.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+I believe the `move` keyword is an excellent candidate for syntax here,
+as it already exists and is also used by the match ergonomics RFC.  
+Alternative keywords would be `const` or `let`.
+Alternatively, a new keyword could be added, although this would need to be a soft keyword or happen over an edition boundary. If this path is desired I suggest `bind`.
```

Comment:
It is written desired **parameter** _property_ currently in sharing part of deconstruct pattern.
Attribute ref mean we wish **parameter** to _be a reference_ or same referential as deconstructed variable if attribute is off.
Attribute mut mean we wish **parameter** to _be a mutable one_ or same changeable as deconstructed variable if attribute is off.

Possible alternative attribute const mean we wish **parameter** to _be a immutable one_ or same changeable as deconstructed variable if attribute is off.

But proposed attribute move mean we wish **deconstruct value** to _be moved_ or same changeable as deconstructed variable if attribute is off.

Reply:
`const` for a binding implies that it is known at compile time, which is not the case here. I also already noted that `const` is an alternative. I’m not sure what you’re trying to say—do you think `const` is a better option? I also do not understand your reasoning. We *are* deconstructing and moving a value. Additionally your language is hard to understand.

Reply:
@schuelermine thank you for adding!
Both `move` & `const`(as `not_mut`) have same strength level of expressiveness.
Simple explanation: `let Some(const x) = &mut y; ` we read as "some of constant(not mutable) **variable X** deconstructed from mutable reference Y"
But   `let Some(move x) = &mut y; ` we read as "some of **moved from variable Y into variable X** deconstructed from mutable reference Y". We don't move any X, but Y.
May be word `new` has a much better meaning, than `move`:
```rust
let Some(new x) = &mut y;
```
But I don't know if this Keyword is possible to make soft.

Reply:
@schuelermine Finally I found the **right word** for _move_-meaning -  `value` !
```rust
let Some(value x) = &mut y;
let Some(ref x)   = &mut y;
```

Reply:
I’ve added a list of possible keywords including your ideas.